ystia/yorc
# Fix attributes notifications for services (substitutions)

# Pull Request description

## Description of the change

Change the way we compute attributes notifications for substituted nodes.

### What I did

* on `deployments/attribute_notifications.go`:
  * register attributes in the hosted-on hierarchy only for nodes that actually have the attribute we are looking for.
  * Factorize the way we are looking for an `ip_address` of a capability derived from `tosca.capabilities.Endpoint`
* on `deployments/capabilities.go`:
  * Refactor and factorize code that allow to select between `private_address` or `public_address` of an host when looking for a `tosca.capabilities.Endpoint` `ip_address`attribute
*  on `deployments/nodes.go`:
  * Added `NodeHasAttribute` & `NodeHasProperty` which are shorthand for getting node type and calling respectively `TypeHasAttribute` and `TypeHasProperty`

### How to verify it

following the How to verify it part of the PR on hybrid demo is a good candidate as it wont work without this https://github.com/ystia/forge/pull/94

### Description for the changelog

* Fix attributes notifications for services (substitutions) ([GH-423](https://github.com/ystia/yorc/issues/423))

## Applicable Issues

Fixes #423
